File: synapsys-benchmarks/src/jmh/java/io/eigr/synapsys/benchmarks/SynapsysBenchmark.java
package io.eigr.synapsys.benchmarks;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.infra.Blackhole;

import java.util.concurrent.TimeUnit;

import io.eigr.synapsys.core.actor.ActorPointer;
import io.eigr.synapsys.core.actor.ActorSystem;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Thread)
public class SynapsysBenchmark {

    private ActorPointer<Message> actor;

    @Setup
    public void setup() {
        ActorSystem system = ActorSystem.INSTANCE;
        system.create();

        actor = system.actorOf(
                "benchmark-actor",
                0,
                null,
                SimpleActor::new
        );

    }
    @Benchmark
    public void benchmarkAsk(Blackhole bh) {
       // bh.consume(actor.ask<String>(new Message("Hello")));
    }

    @Benchmark
    public void benchmarkSend(Blackhole bh) {
        //bh.consume(actor.send(new Message("Hello")));
    }
}

File: synapsys-benchmarks/src/jmh/kotlin/io/eigr/synapsys/benchmarks/SynapsysBenchmark.kt
package io.eigr.synapsys.benchmarks

import io.eigr.synapsys.core.actor.ActorPointer
import io.eigr.synapsys.core.actor.ActorSystem
import kotlinx.coroutines.runBlocking
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.annotations.BenchmarkMode
import org.openjdk.jmh.annotations.Mode
import org.openjdk.jmh.annotations.OutputTimeUnit
import org.openjdk.jmh.annotations.Scope
import org.openjdk.jmh.annotations.Setup
import org.openjdk.jmh.annotations.State
import java.util.concurrent.TimeUnit

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Thread)
open class SynapsysBenchmark {

    private lateinit var actor: ActorPointer<Message>

    @Setup
    fun setup() {
        runBlocking {
            ActorSystem.create()
            actor = ActorSystem.actorOf("benchmark-actor", 0) { id, initialState ->
                SimpleActor(id, initialState)
            }
        }
    }

    @Benchmark
    fun benchmarkAsk(): String {
        return runBlocking { actor.ask<String>(Message("Hello")) }
    }

    @Benchmark
    fun benchmarkSend() {
        runBlocking { actor.send(Message("Hello")) }
    }
}
